### PR TITLE
[WFLY-9004] Artemis leaks JGroups channels on reload

### PR DESCRIPTION
**Pull request by jmesnil**
Refactor messaging-activemq subsystem code to create channels on demand
every time Artemis creates a JGroupsBroadcastEndpoint.

The ActiveMQServerServices track only the channels created by the discovery
groups since they can be referenced by Artemis RA (when the RA is using
JGroups to discover the Artemis servers).

### CODE REVIEW

**Comment by jmesnil** [comment]
@pferraro Could you review this PR please?

I changed the messaging-activemq subsystem so that it creates/closes channel every time that Artemis requests it (instead to have the channels created when the server is started).
Do you see any issue with that new behaviour?
Would there be any difference if the channel is a regular channel or a `ForkChannel`?

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupsBroadcastEndpointFactory.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+import org.jgroups.JChannel;
+import org.wildfly.clustering.jgroups.spi.ChannelFactory;
+
+/**
+ * @author <a href="http://jmesnil.net/">Jeff Mesnil</a> (c) 2017 Red Hat inc.
+ */
+class JGroupsBroadcastEndpointFactory implements BroadcastEndpointFactory {
+    private final String channelName;
+    private final ChannelFactory channelFactory;
+    private final Map<String, JChannel> channels;
```
Comment (pferraro) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupsBroadcastEndpointFactory.java]: Does this map serve any purpose?  Doesn't the JChannelManager already manage channel sharing by multiple endpoints?

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/ActiveMQServerService.java [hunk 2]
```diff
@@ -304,16 +302,11 @@ public synchronized void start(final StartContext context) throws StartException
...
                     if (jgroupFactories.containsKey(key)) {
                         ChannelFactory channelFactory = jgroupFactories.get(key);
                         String channelName = jgroupsChannels.get(key);
-                        JChannel channel = channels.get(channelName);
-                        if (channel == null) {
-                            channel = (JChannel) channelFactory.createChannel(channelName);
-                            channels.put(channelName, channel);
-                        }
-                        config = DiscoveryGroupAdd.createDiscoveryGroupConfiguration(name, entry.getValue(), channel, channelName);
+                        config = DiscoveryGroupAdd.createDiscoveryGroupConfiguration(name, entry.getValue(), channelFactory, channelName, channels);
```
Comment (pferraro) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/ActiveMQServerService.java]: @jmesnil Question - the intention of the previous logic seems to be that only 1 JGroups channel would be created for every unique channel name between the broadcast group and discovery groups.
This change moves the creation of the channel to the endpoint and defers the management of channel sharing between multiple endpoints to the JChannelManager.  I think that makes sense.  Ideally, a JChannel should be closed by the same object that created it.
However, this seems to render the Map<String, JChannel> in ActiveMQServerService obsolete.  Can we remove this?  The channels created by discovery group endpoints are added to this map and removed on close, but I don't see anything that ever reads from this map.

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/BroadcastGroupAdd.java [hunk 3]
```diff
@@ -175,12 +176,12 @@ static BroadcastGroupConfiguration createBroadcastGroupConfiguration(final Strin
...
     }
 
-    static BroadcastGroupConfiguration createBroadcastGroupConfiguration(final String name, final BroadcastGroupConfiguration config, final JChannel channel, final String channelName) throws Exception {
+    static BroadcastGroupConfiguration createBroadcastGroupConfiguration(final String name, final BroadcastGroupConfiguration config, final ChannelFactory channelFactory, final String channelName) throws Exception {
 
         final long broadcastPeriod = config.getBroadcastPeriod();
         final List<String> connectorRefs = config.getConnectorInfos();
 
-        final BroadcastEndpointFactory endpointFactory = new ClosingChannelBroadcastEndpointFactory(channel, channelName);
+        final BroadcastEndpointFactory endpointFactory = new JGroupsBroadcastEndpointFactory(channelName, channelFactory, new HashMap<>());
```
Comment (pferraro) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/BroadcastGroupAdd.java]: This map seems to serve no purpose other than to make JGroupsBroadcastEndpointFactory happy.

**Review by pferraro — CHANGES_REQUESTED** [review]

**Comment by pferraro** [comment]
@jmesnil In general, I think this change makes sense.  I like that the same object that creates the channel closes it.  In the case where the ChannelFactory is a ForkChannelFactory, JChannel will still be created on startup (a ForkChannelFactory is built on top of a JChannel), but the fork channels will not be created until ActiveMQ request them from the endpoint.

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupsBroadcastEndpointFactory.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+                synchronized (JGroupsBroadcastEndpointFactory.class) {
+                    JChannel channel = (JChannel) channelFactory.createChannel(channelName);
+                    channels.put(channelName, channel);
+                    return channel;
+                }
+            }
+
+            @Override
+            protected synchronized void internalCloseChannel(JChannelWrapper channel) {
+                synchronized (JGroupsBroadcastEndpointFactory.class) {
```
Comment (pferraro) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupsBroadcastEndpointFactory.java]: Why the funky synchronization on both the instance and the parent class?

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupsBroadcastEndpointFactory.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        this.channelFactory = channelFactory;
+        this.channels = channels;
+    }
+
+    @Override
+    public BroadcastEndpoint createBroadcastEndpoint() throws Exception {
+        JGroupsBroadcastEndpoint endpoint = new JGroupsBroadcastEndpoint(JChannelManager.getInstance(), channelName) {
+            @Override
+            public JChannel createChannel() throws Exception {
+                synchronized (JGroupsBroadcastEndpointFactory.class) {
```
Comment (pferraro) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupsBroadcastEndpointFactory.java]: Why is this synchronized on the class?  Does this need to be synchronized at all?

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/ActiveMQServerService.java (hunk 2 above)
Comment (jmesnil) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/ActiveMQServerService.java]: You are right about the intent of the previous logic. However I was not matching Artemis expectations (which creates channels as it needs it without sharing them between its broadcast and discovery groups).

I'm not sure there is a correct use case for sharing the same channel name between discovery groups. The discovery groups would only differ by their `refresh-timeout` and `initial-wait-timeout`attributes. I don't see a compelling use case for this but it's technically acceptable...

**Comment by jmesnil** [comment]
@pferraro thanks for the review. I'll change the `JGroupsBroadcastEndpointFactory` to get rid of that ugly empty Map to make it more readable that only channels created by discovery groups needs to be accessed by the Artemis RA.

**Review by pferraro — CHANGES_REQUESTED** [review]

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupsBroadcastEndpointFactory.java (hunk 5 above)
Comment (jmesnil) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupsBroadcastEndpointFactory.java]: fixed

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupsBroadcastEndpointFactory.java (hunk 4 above)
Comment (jmesnil) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupsBroadcastEndpointFactory.java]: fixed

**Pull request merged**